### PR TITLE
Proposal: Federating namespace restrictions.

### PR DESCRIPTION
**Pull request by mattmoor**
See: registry/federated-namespaces.md

Signed-off-by: Matt Moore mattmoor@google.com

### CODE REVIEW

**Comment by mattmoor** [comment]
@dmp42 I wrote up this proposal to better capture my offhand remark from https://github.com/docker/distribution/pull/483

Please LMK what you think, and who should take a look at this.

File: registry/federated-namespaces.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# Proposal: Federating namespace restrictions
+
+This proposes minimizing the constraints the Docker client places on image names, deferring instead to the Registry at push/pull time to push back with a suitable error message around character restrictions / form.
+
+
+## Preface: Current State
+
+Today, the client is overly aware of the character restrictions and form of repository names.  These restrictions have been adjusted throughout the course of v1, and have changed non-trivially in the v1 => v2 switch.
+
+The restrictions enforced in the client today are very v1-centric, which
```
Comment (mattdr) [inline registry/federated-namespaces.md]: unfinished sentence

File: registry/federated-namespaces.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+The restrictions enforced in the client today are very v1-centric, which 
+
+### v1
+
+In the world of Docker Registry v1, names have the form:
+```
+   registry/namespace/image:tag
+```
+
+```tag``` can often be omitted, and depending on context can mean ```latest``` or all tags.
```
Comment (mattdr) [inline registry/federated-namespaces.md]: You can use single backticks inline.

Comment (mattmoor) [inline registry/federated-namespaces.md]: Done :)

File: registry/federated-namespaces.md (hunk 1 above)
Comment (mattmoor) [inline registry/federated-namespaces.md]: Oops, should be fixed.  thanks,

**Comment by mattmoor** [comment]
Thanks @mattdr should be fixed.

**Comment by mattmoor** [comment]
@dmp42 ping?

File: registry/federated-namespaces.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+In order to continue to support both v1 and v2 within the same client, while enabling customers to start leveraging the more flexible aspects of naming afforded by the v2 registry, I propose the following grammar for image paths:
+```
+//  tag := /[^:]+/
+//  hash := 'sha256'
+//        | 'sha284'
+//        | 'sha512'
+//        | 'tarsum'     // TODO(mattmoor): Can we retire this?
+//  digest := /[a-fA-F0-9]+/
+//  image := /.+/
```
Comment (dmp42) [inline registry/federated-namespaces.md]: This (probably) doesn't work for the reasons mentioned earlier.
Also, remember that tagging is an entirely client-side operation. Letting people tag freely, only to see their push failing down the road may be a questionable approach. Also, I'm concerned about moving validation to the server, for similar reasons: a given image name might work with registry A, and not with registry B.

**Comment by dmp42** [comment]
ping @tiborvass @dmcgowan (+ @stevvooe and @RichardScothern)

File: registry/federated-namespaces.md (hunk 3 above)
Comment (mattmoor) [inline registry/federated-namespaces.md]: FWIW, by making the v2-naming space a non-superset of v1, you have already created this situation.  In fact, I have included an example where I do precisely this with the 1.6 client towards the end.

Moreover, you cannot possibly allow folks to *start* using the expanded namespace afforded by v2 without considering the case where it fails when pushing to v1.

File: registry/federated-namespaces.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+In the world of Docker Registry v2, names have a similar, but slightly more flexible form best described [here](https://sourcegraph.com/github.com/docker/distribution@5556cd1ba11aea6803f3c51a16bb96813ba72800/.tree/registry/api/v2/names.go#selected=56):
+```
+// Effectively, the name should comply with the following grammar:
+//
+// 	alpha-numeric := /[a-z0-9]+/
+//	separator := /[._-]/
+//	component := alpha-numeric [separator alpha-numeric]*
+//	namespace := component ['/' component]*
+```
+Where `namespace` is used in the context: `registry/namespace:tag`, where as before `registry` and `tag` are optional.
```
Comment (stevvooe) [inline registry/federated-namespaces.md]: V2 does not prescribed this definition of namespace. This description is more of the conventions used on the hub. The number of components is not restricted and the meaning of each component is use case specific.

**Comment by stevvooe** [comment]
@mattmoor I think you'll get a lot further in this regard if you work through the docker/distribution project where we are actively discussing these issues. Specifically, this is directly relevant to https://github.com/docker/distribution/pull/303.

Regarding the actual validation of names, the only path forward is to make the docker client and the registry consistent. There is no good reason to segment the community by registries that support different naming conventions. In other words, we will relax naming to the V2 API specification but no further. (The fact that the docker daemon allows tagging images as "mattmoor/_blazinga" is a bug).

Arguably, we are missing a few specifications for the grammar of names, such as digests. A good starting point might be to add those to the V2 API specification.

**Comment by dmp42** [comment]
@stevvooe they filed here on my suggestion (though admittedly the question was specifically about lifting the client restriction on colons in name portions).

@mattmoor apology for making you move around repositories - hopefully @stevvooe indications will better help you.

My remark still stands about colons though - creating ambiguity for the user regarding separation between name and tag is probably not going to fly.

**Comment by mattmoor** [comment]
@stevvooe I added some commentary on the linked issue.  Let's close this and continue the discussion there, but I'm hoping that through a discovery service we might be able to achieve both of our goals.

Closing this for now.

**Pull request closed without merging**